Add LD2410 BLE integration

930913 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Adds an MVP integration to connect to LD2410B sensors over bluetooth. This was the mmWave sensor demonstrated by Andreas Spiess at https://www.youtube.com/watch?v=dAzHXpP3FcI

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/25264

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [n/a] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

930913 committed:
Add LD2410 BLE integration

- Add integration
- Add config flow
- Add binary_sensor
- Add tests
- Update CODEOWNERS

930913 committed:
Fix pylint

- Add return types
- Raise errors from previous error
- Rm unneeded log

930913 committed:
Fix types

Add type notation to missing places in binary_sensor

930913 committed:
Fix typos

update -> initialise

bdraco commented on homeassistant/components/ld2410_ble/binary_sensor.py:
```diff
@@ -0,0 +1,112 @@
+from . import LD2410BLE
+from .const import DOMAIN
+from .models import LD2410BLEData
+
```
Use BinarySensorEntityDescription here instead

930913 replied:
Done.

bdraco commented on homeassistant/components/ld2410_ble/binary_sensor.py:
```diff
@@ -0,0 +1,112 @@
+) -> None:
+    """Set up the platform for LD2410BLE."""
+    data: LD2410BLEData = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities([IsMovingSensor(data.coordinator, data.device, entry.title)])
+    async_add_entities([IsStaticSensor(data.coordinator, data.device, entry.title)])
```
Please only call `async_add_entities` once since each creates a task

930913 replied:
Done.

bdraco commented on homeassistant/components/ld2410_ble/translations/ca.json:
```diff
@@ -0,0 +1,23 @@
+{
+    "config": {
+        "abort": {
+            "already_configured": "El dispositiu ja est\u00e0 configurat",
+            "already_in_progress": "El flux de configuraci\u00f3 ja est\u00e0 en curs",
+            "no_devices_found": "No s'han trobat dispositius a la xarxa",
+            "no_unconfigured_devices": "No s'han trobat dispositius no configurats.",
+            "not_supported": "Dispositiu no compatible"
+        },
+        "error": {
+            "cannot_connect": "Ha fallat la connexi\u00f3",
```
Drop any languages except en.json since they will get overwritten by lokalize anyways

930913 replied:
That's neat. Done.

930913 committed:
Use BinarySensorEntityDescription

- Use BinarySensorEntityDescription
- Remove lokalized languages
- Remove dupe async_add_entities call

bdraco commented on homeassistant/components/ld2410_ble/__init__.py:
```diff
@@ -0,0 +1,119 @@
+
+    try:
+        await ld2410_ble.initialise()
+    except BleakNotFoundError as exc:
```
You should probably trap any `BleakError` here

930913 replied:
Good idea. Done.

bdraco commented on homeassistant/components/ld2410_ble/binary_sensor.py:
```diff
@@ -0,0 +1,116 @@
+    data: LD2410BLEData = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities(
+        [
+            IsMovingSensor(data.coordinator, data.device, entry.title),
+            IsStaticSensor(data.coordinator, data.device, entry.title),
```
Pass in the description and set `self.entity_description` instead so you only need one class

930913 replied:
Updated, but I'm not convinced having a single class makes it more readable. It took me quite a while to unpick other examples in order to understand how to do this stuff

bdraco commented on homeassistant/components/ld2410_ble/binary_sensor.py:
```diff
@@ -0,0 +1,116 @@
+        self._value = self._device.is_static
+        self.async_write_ha_state()
+
+    @property
+    def should_poll(self) -> bool:
+        """Don't poll."""
+        return False
+
```
CoordinatorEntity already sets this

930913 replied:
That's cool. Removed.

930913 committed:
Refactor to one class

Also catch all BleakError

bdraco commented on homeassistant/components/ld2410_ble/binary_sensor.py:
```diff
@@ -0,0 +1,75 @@
+        self._key = key
+        self._device = device
+        self.entity_description = ENTITY_DESCRIPTIONS[key]
+        self._attr_unique_id = device.address + f"_{key}"
```
```suggestion
        self._attr_unique_id = f"{device.address}_{key}"
```

930913 replied:
Good spot.

930913 committed:
Comprehend sensors

And don't concatenate f strings.

930913 commented on homeassistant/components/ld2410_ble/binary_sensor.py:
```diff
@@ -0,0 +1,75 @@
+            name=name,
+            connections={(dr.CONNECTION_BLUETOOTH, device.address)},
+        )
+        self._attr_is_on = False
```
It should probably be unavailable until read from the device, but the library will populate it to False initially. Changed to this anyhow.

930913 commented on homeassistant/components/ld2410_ble/__init__.py:
```diff
@@ -0,0 +1,119 @@
+
+    def _async_handle_disconnect(self) -> None:
+        """Trigger the callbacks with False for disconnected."""
+        self.async_set_update_error(
+            Exception(f"Disconnected from {self._ld2410_ble.name}")
+        )
```
This isn't really a bluetooth or device exception. `async_set_update_error` takes an exception as a parameter, but in the disconnect callback (which is not exceptional) there is no exception, so I constructed this to pacify the call.

bdraco replied:
You could have the lib pass back a `LD2410DisconnectedError` via the `register_disconnected_callback` and pass that back to `async_set_update_error` so if you ever wanted change the message or raise a different exception based on why the disconnect happened it would be a lib update and no changes in core

930913 replied:
There's a bit of a code smell there with the tail wagging the dog. The library never had an exception, so why should it construct one for this use case? At best, it should perhaps pass the disconnect reason to the callback, but the exception still needs to be constructed from it at the point of the `async_set_update_error` call.

bdraco replied:
You could instead get rid of the call to `async_set_update_error`, call `async_update_listeners` instead, add a new property to the coordinator called `connected` and override `available` in the entity to check `super().available and self.connected`

930913 replied:
That's what I had originally written, but I couldn't get it working. I just rewrote it back in, and it now works... ¯\\\_(ツ)\_/¯

bdraco commented on homeassistant/components/ld2410_ble/binary_sensor.py:
```diff
@@ -0,0 +1,71 @@
+from .const import DOMAIN
+from .models import LD2410BLEData
+
+ENTITY_DESCRIPTIONS = {
+    "is_moving": BinarySensorEntityDescription(
+        key="is_moving", device_class=BinarySensorDeviceClass.MOTION
+    ),
+    "is_static": BinarySensorEntityDescription(
+        key="is_static", device_class=BinarySensorDeviceClass.PRESENCE
+    ),
+}
```
```suggestion
ENTITY_DESCRIPTIONS = {
    BinarySensorEntityDescription(
        key="is_moving", device_class=BinarySensorDeviceClass.MOTION
    ),
    BinarySensorEntityDescription(
        key="is_static", device_class=BinarySensorDeviceClass.PRESENCE
    ),
}
```

930913 replied:
Updated.

bdraco commented on homeassistant/components/ld2410_ble/binary_sensor.py:
```diff
@@ -0,0 +1,71 @@
+    data: LD2410BLEData = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities(
+        [
+            LD2410BLEBinarySensor(data.coordinator, data.device, entry.title, k)
+            for k in ENTITY_DESCRIPTIONS
```
```suggestion
            LD2410BLEBinarySensor(data.coordinator, data.device, entry.title, description)
            for description in ENTITY_DESCRIPTIONS
```

930913 replied:
Done.

bdraco commented on homeassistant/components/ld2410_ble/binary_sensor.py:
```diff
@@ -0,0 +1,71 @@
+        self._coordinator = coordinator
+        self._key = key
+        self._device = device
+        self.entity_description = ENTITY_DESCRIPTIONS[key]
```
```suggestion
        self.entity_description = description
```

930913 replied:
Done.

bdraco commented on homeassistant/components/ld2410_ble/binary_sensor.py:
```diff
@@ -0,0 +1,71 @@
+        """Initialize the sensor."""
+        super().__init__(coordinator)
+        self._coordinator = coordinator
+        self._key = key
```
```suggestion
        self._key = description.key
```

930913 replied:
Kept, so as not to traverse entity_description when retrieving, which I think makes the code easier to understand.

bdraco commented on homeassistant/components/ld2410_ble/binary_sensor.py:
```diff
@@ -0,0 +1,71 @@
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        self._attr_is_on = getattr(self._device, self._key)
```
```suggestion
        self._attr_is_on = getattr(self._device, self.description.key)
```

930913 replied:
Not changed. (See other comment.)

bdraco commented on homeassistant/components/ld2410_ble/binary_sensor.py:
```diff
@@ -0,0 +1,71 @@
+            name=name,
+            connections={(dr.CONNECTION_BLUETOOTH, device.address)},
+        )
+        self._attr_is_on = getattr(self._device, self._key)
```
```suggestion
        self._attr_is_on = getattr(self._device, self.description.key)
```

930913 replied:
Not changed. (See other comment.)

bdraco commented on homeassistant/components/ld2410_ble/binary_sensor.py:
```diff
@@ -0,0 +1,71 @@
+    """Set up the platform for LD2410BLE."""
+    data: LD2410BLEData = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities(
+        [
```
You can drop the `[]` as well as a generator expression works here

930913 replied:
Done.

bdraco commented on homeassistant/components/ld2410_ble/binary_sensor.py:
```diff
@@ -0,0 +1,71 @@
+from .const import DOMAIN
+from .models import LD2410BLEData
+
+ENTITY_DESCRIPTIONS = {
```
https://developers.home-assistant.io/docs/core/entity?_highlight=has_entit#entity-naming

has_entity_name True (Mandatory for new integrations)

930913 replied:
Added.

bdraco commented on homeassistant/components/ld2410_ble/const.py:
```diff
@@ -0,0 +1,9 @@
+DEVICE_TIMEOUT = 30
+LOCAL_NAMES = {"HLK-LD2410B"}
+
+UPDATE_SECONDS = 15
```
This doesn't look like its used

930913 replied:
The linter isn't as powerful as I assumed. Removed.

930913 committed:
Fix manifest.documentation

s/led/ld2410/

930913 committed:
Remove unused constants

930913 committed:
Pack entity description

Use the entity description to get the key

930913 committed:
Refactor availibility

Track connectivity on the coordinator and return availiblity accordingly.

930913 committed:
Add coordinator file

Forgot to add it, lol.

930913 commented on homeassistant/components/ld2410_ble/binary_sensor.py:
```diff
@@ -0,0 +1,71 @@
+    """Moving/static sensor for LD2410BLE."""
+
+    def __init__(
+        self, coordinator: DataUpdateCoordinator, device: LD2410BLE, name: str, key: str
```
Done.

930913 commented on homeassistant/components/ld2410_ble/const.py:
```diff
@@ -0,0 +1,9 @@
+
+DOMAIN = "ld2410_ble"
+
+DEVICE_TIMEOUT = 30
```
The linter isn't as powerful as I assumed. Removed.

930913 commented on homeassistant/components/ld2410_ble/manifest.json:
```diff
@@ -0,0 +1,12 @@
+  "domain": "ld2410_ble",
+  "name": "LD2410 BLE",
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/led_ble/",
```
Must have copied from an old version when recovering my laptop, lol. Fixed.

bdraco commented on homeassistant/components/ld2410_ble/manifest.json:
```diff
@@ -0,0 +1,12 @@
+  "requirements": ["bluetooth-data-tools==0.3.0", "ld2410-ble==0.1.1"],
+  "dependencies": ["bluetooth"],
+  "codeowners": ["@930913"],
+  "bluetooth": [{ "local_name": "HLK-LD2410B_*" }],
```
Is there another way to detect it without using the long name since unless it broadcasts this it won't be able to be found with passive scans

930913 replied:
After much trying, I can inconclusively say no: the only way it showed up as discovered was with the name. Could it just be my BT adaptor?

bdraco replied:
It's probably just a limitation of the device itself. That's unfortunate

930913 committed:
Add coordinator file to coverage exclusion

TODO: Tests

bdraco commented:
This looks like its pretty close to ready.  I just need to find time to do another review pass after I've had enough sleep to be sure 😄

930913 commented:
> This looks like its pretty close to ready. I just need to find time to do another review pass after I've had enough sleep to be sure smile

@bdraco How has your sleeping been? :laughing:

bdraco commented:
Not as well as I had hoped.  This is still on my list to get done

bdraco commented on homeassistant/components/ld2410_ble/binary_sensor.py:
```diff
@@ -0,0 +1,81 @@
+from .const import DOMAIN
+from .models import LD2410BLEData
+
+ENTITY_DESCRIPTIONS = [
+    BinarySensorEntityDescription(
+        key="is_moving",
+        device_class=BinarySensorDeviceClass.MOTION,
+        has_entity_name=True,
+        name="Motion",
+    ),
+    BinarySensorEntityDescription(
+        key="is_static",
+        device_class=BinarySensorDeviceClass.OCCUPANCY,
+        has_entity_name=True,
+        name="Occupancy",
+    ),
+]
```
MINOR: This could be a tuple since it never changes

930913 replied:
Noted. I might change it in a subsequent MR.

bdraco reviewed (approved):
LGTM 👍

bdraco commented:
Looks like there is some feedback on the docs PR to address: https://github.com/home-assistant/home-assistant.io/pull/25264
